File: frontend/src/services/tenantAuthService.ts
import { api } from '@/utils/api';

export interface TenantAuthConfig {
  id: string;
  company_id: string;
  custom_login_url?: string;
  company_code_required: boolean;
  session_timeout_minutes: number;
  remember_me_enabled: boolean;
  remember_me_duration_days: number;
  concurrent_sessions_limit: number;
  password_reset_enabled: boolean;
  password_reset_expiry_hours: number;
  custom_reset_template?: string;
  oauth_providers?: Record<string, any>;
  saml_config?: Record<string, any>;
  login_logo_url?: string;
  login_background_url?: string;
  brand_colors?: Record<string, string>;
  created_at: string;
  updated_at: string;
}

export interface TenantLoginRequest {
  email: string;
  password: string;
  company_code?: string;
  remember_me?: boolean;
}

export interface TenantLoginResponse {
  access_token: string;
  token_type: string;
  expires_in: number;
  user_id: string;
  company_id: string;
  company_name: string;
  redirect_url?: string;
}

export interface TenantSession {
  id: string;
  session_token: string;
  user_id: string;
  company_id: string;
  login_method: string;
  ip_address?: string;
  user_agent?: string;
  expires_at: string;
  last_activity: string;
  remember_me: boolean;
  status: string;
  created_at: string;
}

export interface OAuthProvider {
  id: string;
  company_id: string;
  provider_name: string;
  client_id: string;
  redirect_uri: string;
  scopes?: string[];
  is_active: boolean;
  auto_create_users: boolean;
  default_role?: string;
  created_at: string;
}

/**
 * Tenant Authentication Service
 * Provides methods for multi-tenant authentication operations
 */
export default {
  /**
   * Create authentication configuration for a company
   * @param companyId - Company ID
   * @param config - Auth configuration
   * @returns Promise with created config
   */
  async createAuthConfig(companyId: string, config: Partial<TenantAuthConfig>) {
    return api.post(`/tenant-auth/config/${companyId}`, config);
  },

  /**
   * Get authentication configuration for a company
   * @param companyId - Company ID
   * @returns Promise with auth config
   */
  async getAuthConfig(companyId: string) {
    return api.get(`/tenant-auth/config/${companyId}`);
  },

  /**
   * Login with tenant context
   * @param loginData - Login credentials
   * @returns Promise with login response
   */
  async login(loginData: TenantLoginRequest) {
    return api.post('/tenant-auth/login', loginData);
  },

  /**
   * Logout and terminate session
   * @param sessionToken - Session token
   * @returns Promise with logout response
   */
  async logout(sessionToken: string) {
    return api.post('/tenant-auth/logout', { session_token: sessionToken });
  },

  /**
   * Get active sessions for a user
   * @param userId - User ID
   * @param companyId - Company ID
   * @returns Promise with sessions list
   */
  async getUserSessions(userId: string, companyId: string) {
    return api.get(`/tenant-auth/sessions/${userId}?company_id=${companyId}`);
  },

  /**
   * Terminate all sessions for a user
   * @param userId - User ID
   * @param companyId - Company ID
   * @param exceptCurrent - Keep current session active
   * @returns Promise with termination response
   */
  async terminateAllSessions(userId: string, companyId: string, exceptCurrent: boolean = true) {
    return api.delete(`/tenant-auth/sessions/${userId}/terminate-all?company_id=${companyId}&except_current=${exceptCurrent}`);
  },

  /**
   * Request password reset
   * @param email - User email
   * @param companyCode - Company code (optional)
   * @returns Promise with reset response
   */
  async requestPasswordReset(email: string, companyCode?: string) {
    return api.post('/tenant-auth/password-reset', {
      email,
      company_code: companyCode
    });
  },

  /**
   * Confirm password reset
   * @param token - Reset token
   * @param newPassword - New password
   * @returns Promise with confirmation response
   */
  async confirmPasswordReset(token: string, newPassword: string) {
    return api.post('/tenant-auth/password-reset/confirm', {
      token,
      new_password: newPassword
    });
  },

  /**
   * Create OAuth provider configuration
   * @param companyId - Company ID
   * @param providerData - OAuth provider data
   * @returns Promise with created provider
   */
  async createOAuthProvider(companyId: string, providerData: Partial<OAuthProvider>) {
    return api.post(`/tenant-auth/oauth/${companyId}`, providerData);
  },

  /**
   * Get OAuth providers for a company
   * @param companyId - Company ID
   * @param activeOnly - Show only active providers
   * @returns Promise with providers list
   */
  async getOAuthProviders(companyId: string, activeOnly: boolean = true) {
    return api.get(`/tenant-auth/oauth/${companyId}?active_only=${activeOnly}`);
  },

  /**
   * Get companies available to a user
   * @param userId - User ID
   * @returns Promise with companies list
   */
  async getUserCompanies(userId: string) {
    return api.get(`/tenant-auth/companies/${userId}`);
  },

  /**
   * Get login attempts for monitoring
   * @param companyId - Company ID (optional)
   * @param email - Email filter (optional)
   * @param limit - Results limit
   * @returns Promise with login attempts
   */
  async getLoginAttempts(companyId?: string, email?: string, limit: number = 100) {
    const params = new URLSearchParams();
    if (companyId) params.append('company_id', companyId);
    if (email) params.append('email', email);
    params.append('limit', limit.toString());

    return api.get(`/tenant-auth/login-attempts?${params.toString()}`);
  },

  /**
   * Clean up expired sessions
   * @returns Promise with cleanup response
   */
  async cleanupExpiredSessions() {
    return api.post('/tenant-auth/cleanup-sessions');
  },

  /**
   * Utility functions for tenant authentication
   */
  utils: {
    /**
     * Get login URL for a company
     */
    getLoginUrl(companyCode?: string): string {
      const baseUrl = '/login';
      return companyCode ? `${baseUrl}?company=${companyCode}` : baseUrl;
    },

    /**
     * Format session status
     */
    formatSessionStatus(status: string): string {
      const statusMap: Record<string, string> = {
        'active': 'Active',
        'expired': 'Expired',
        'terminated': 'Terminated'
      };
      
      return statusMap[status] || status.charAt(0).toUpperCase() + status.slice(1);
    },

    /**
     * Get status color for UI
     */
    getStatusColor(status: string): string {
      const colorMap: Record<string, string> = {
        'active': 'success',
        'expired': 'warning',
        'terminated': 'error'
      };
      
      return colorMap[status] || 'grey';
    },

    /**
     * Format login method
     */
    formatLoginMethod(method: string): string {
      const methodMap: Record<string, string> = {
        'email_password': 'Email & Password',
        'oauth_google': 'Google OAuth',
        'oauth_microsoft': 'Microsoft OAuth',
        'saml_sso': 'SAML SSO'
      };
      
      return methodMap[method] || method.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase());
    },

    /**
     * Check if session is expiring soon
     */
    isSessionExpiringSoon(expiresAt: string, thresholdMinutes: number = 5): boolean {
      const expiryTime = new Date(expiresAt).getTime();
      const now = new Date().getTime();
      const threshold = thresholdMinutes * 60 * 1000;
      
      return (expiryTime - now) <= threshold;
    },

    /**
     * Get time until session expires
     */
    getTimeUntilExpiry(expiresAt: string): string {
      const expiryTime = new Date(expiresAt).getTime();
      const now = new Date().getTime();
      const diff = expiryTime - now;
      
      if (diff <= 0) return 'Expired';
      
      const minutes = Math.floor(diff / (1000 * 60));
      const hours = Math.floor(minutes / 60);
      const days = Math.floor(hours / 24);
      
      if (days > 0) return `${days}d ${hours % 24}h`;
      if (hours > 0) return `${hours}h ${minutes % 60}m`;
      return `${minutes}m`;
    },

    /**
     * Validate company code format
     */
    validateCompanyCode(code: string): { valid: boolean; message?: string } {
      if (!code) {
        return { valid: false, message: 'Company code is required' };
      }
      
      if (code.length < 2 || code.length > 20) {
        return { valid: false, message: 'Company code must be 2-20 characters' };
      }
      
      if (!/^[A-Za-z0-9-_]+$/.test(code)) {
        return { valid: false, message: 'Company code can only contain letters, numbers, hyphens, and underscores' };
      }
      
      return { valid: true };
    },

    /**
     * Get OAuth provider icon
     */
    getOAuthProviderIcon(provider: string): string {
      const iconMap: Record<string, string> = {
        'google': 'mdi-google',
        'microsoft': 'mdi-microsoft',
        'github': 'mdi-github',
        'linkedin': 'mdi-linkedin'
      };
      
      return iconMap[provider.toLowerCase()] || 'mdi-account-circle';
    },

    /**
     * Generate secure session token
     */
    generateSessionToken(): string {
      const array = new Uint8Array(32);
      crypto.getRandomValues(array);
      return Array.from(array, byte => byte.toString(16).padStart(2, '0')).join('');
    }
  }
};

File: backend/app/services/auth/tenant_auth_service.py
"""
Multi-tenant authentication service.
"""
from datetime import datetime, timedelta
from typing import Optional, Dict, Any, List

from sqlalchemy import and_, desc
from sqlalchemy.orm import Session
from uuid import UUID
import secrets

from app.core.security.encryption import encrypt_data, decrypt_data
from app.models.tenant_auth import (



    TenantAuthConfig, TenantSession, CompanyLoginAttempt, 
    PasswordResetToken, OAuthProvider, LoginMethod, SessionStatus
)


class TenantAuthService:
    """Service for multi-tenant authentication operations."""
    
    def __init__(self, db: Session):
        """  Init  ."""
        self.db = db
    
    def get_auth_config(self, company_id: UUID) -> Optional[TenantAuthConfig]:
        """Get Auth Config."""
        """Get authentication configuration for a company."""
        return self.db.query(TenantAuthConfig).filter(
            TenantAuthConfig.company_id == company_id
        ).first()
    
    def create_auth_config(self, company_id: UUID, config_data: Dict[str, Any]) -> TenantAuthConfig:
        """Create Auth Config."""
        """Create authentication configuration for a company."""
        config = TenantAuthConfig(
            company_id=company_id,
            custom_login_url=config_data.get('custom_login_url'),
            company_code_required=config_data.get('company_code_required', False),
            session_timeout_minutes=config_data.get('session_timeout_minutes', 30),
            remember_me_enabled=config_data.get('remember_me_enabled', True),
            remember_me_duration_days=config_data.get('remember_me_duration_days', 30),
            concurrent_sessions_limit=config_data.get('concurrent_sessions_limit', 5),
            password_reset_enabled=config_data.get('password_reset_enabled', True),
            password_reset_expiry_hours=config_data.get('password_reset_expiry_hours', 24),
            custom_reset_template=config_data.get('custom_reset_template'),
            oauth_providers=config_data.get('oauth_providers'),
            saml_config=config_data.get('saml_config'),
            login_logo_url=config_data.get('login_logo_url'),
            login_background_url=config_data.get('login_background_url'),
            brand_colors=config_data.get('brand_colors')
        )
        
        self.db.add(config)
        self.db.commit()
        self.db.refresh(config)
        
        return config
    
    def create_session(
        self,
        user_id: UUID,
        company_id: UUID,
        login_method: str = LoginMethod.EMAIL_PASSWORD,
        ip_address: Optional[str] = None,
        user_agent: Optional[str] = None,
        remember_me: bool = False
    ) -> TenantSession:
        """Create Session."""
        """Create a new tenant session."""
        # Get auth config for session timeout
        auth_config = self.get_auth_config(company_id)
        timeout_minutes = auth_config.session_timeout_minutes if auth_config else 30
        
        # Calculate expiry
        if remember_me and auth_config and auth_config.remember_me_enabled:
            expires_at = datetime.utcnow() + timedelta(days=auth_config.remember_me_duration_days)
        else:
            expires_at = datetime.utcnow() + timedelta(minutes=timeout_minutes)
        
        # Check concurrent sessions limit
        if auth_config:
            self._enforce_session_limit(user_id, company_id, auth_config.concurrent_sessions_limit)
        
        # Generate session token
        session_token = secrets.token_urlsafe(32)
        
        session = TenantSession(
            session_token=session_token,
            user_id=user_id,
            company_id=company_id,
            login_method=login_method,
            ip_address=ip_address,
            user_agent=user_agent,
            expires_at=expires_at,
            remember_me=remember_me,
            status=SessionStatus.ACTIVE
        )
        
        self.db.add(session)
        self.db.commit()
        self.db.refresh(session)
        
        return session
    
    def validate_session(self, session_token: str) -> Optional[TenantSession]:
        """Validate Session."""
        """Validate and refresh a session."""
        session = self.db.query(TenantSession).filter(
            and_(
                TenantSession.session_token == session_token,
                TenantSession.status == SessionStatus.ACTIVE,
                TenantSession.expires_at > datetime.utcnow()
            )
        ).first()
        
        if session:
            # Update last activity
            session.last_activity = datetime.utcnow()
            
            # Extend expiry if not remember_me session
            if not session.remember_me:
                auth_config = self.get_auth_config(session.company_id)
                timeout_minutes = auth_config.session_timeout_minutes if auth_config else 30
                session.expires_at = datetime.utcnow() + timedelta(minutes=timeout_minutes)
            
            self.db.commit()
        
        return session
    
    def terminate_session(self, session_token: str, reason: str = "user_logout") -> bool:
        """Terminate Session."""
        """Terminate a session."""
        session = self.db.query(TenantSession).filter(
            TenantSession.session_token == session_token
        ).first()
        
        if session:
            session.status = SessionStatus.TERMINATED
            session.terminated_reason = reason
            self.db.commit()
            return True
        
        return False
    
    def terminate_all_user_sessions(self, user_id: UUID, company_id: UUID, except_token: Optional[str] = None):
        """Terminate All User Sessions."""
        """Terminate all sessions for a user in a company."""
        query = self.db.query(TenantSession).filter(
            and_(
                TenantSession.user_id == user_id,
                TenantSession.company_id == company_id,
                TenantSession.status == SessionStatus.ACTIVE
            )
        )
        
        if except_token:
            query = query.filter(TenantSession.session_token != except_token)
        
        sessions = query.all()
        
        for session in sessions:
            session.status = SessionStatus.TERMINATED
            session.terminated_reason = "admin_termination"
        
        self.db.commit()
    
    def log_login_attempt(
        self,
        email: str,
        success: bool,
        company_id: Optional[UUID] = None,
        login_method: str = LoginMethod.EMAIL_PASSWORD,
        failure_reason: Optional[str] = None,
        ip_address: Optional[str] = None,
        user_agent: Optional[str] = None
    ) -> CompanyLoginAttempt:
        """Log Login Attempt."""
        """Log a login attempt."""
        attempt = CompanyLoginAttempt(
            company_id=company_id,
            email=email,
            success=success,
            login_method=login_method,
            failure_reason=failure_reason,
            ip_address=ip_address,
            user_agent=user_agent
        )
        
        self.db.add(attempt)
        self.db.commit()
        self.db.refresh(attempt)
        
        return attempt
    
    def create_password_reset_token(
        self,
        user_id: UUID,
        company_id: UUID,
        email: str
    ) -> PasswordResetToken:
        """Create Password Reset Token."""
        """Create a password reset token."""
        # Invalidate existing tokens
        self.db.query(PasswordResetToken).filter(
            and_(
                PasswordResetToken.user_id == user_id,
                PasswordResetToken.company_id == company_id,
                PasswordResetToken.used == False
            )
        ).update({"used": True, "used_at": datetime.utcnow()})
        
        # Get auth config for expiry
        auth_config = self.get_auth_config(company_id)
        expiry_hours = auth_config.password_reset_expiry_hours if auth_config else 24
        
        token = secrets.token_urlsafe(32)
        
        reset_token = PasswordResetToken(
            token=token,
            user_id=user_id,
            company_id=company_id,
            email=email,
            expires_at=datetime.utcnow() + timedelta(hours=expiry_hours)
        )
        
        self.db.add(reset_token)
        self.db.commit()
        self.db.refresh(reset_token)
        
        return reset_token
    
    def validate_password_reset_token(self, token: str) -> Optional[PasswordResetToken]:
        """Validate Password Reset Token."""
        """Validate a password reset token."""
        return self.db.query(PasswordResetToken).filter(
            and_(
                PasswordResetToken.token == token,
                PasswordResetToken.used == False,
                PasswordResetToken.expires_at > datetime.utcnow()
            )
        ).first()
    
    def use_password_reset_token(self, token: str) -> bool:
        """Use Password Reset Token."""
        """Mark a password reset token as used."""
        reset_token = self.validate_password_reset_token(token)
        
        if reset_token:
            reset_token.used = True
            reset_token.used_at = datetime.utcnow()
            self.db.commit()
            return True
        
        return False
    
    def create_oauth_provider(
        self,
        company_id: UUID,
        provider_data: Dict[str, Any]
    ) -> OAuthProvider:
        """Create Oauth Provider."""
        """Create OAuth provider configuration."""
        provider = OAuthProvider(
            company_id=company_id,
            provider_name=provider_data['provider_name'],
            client_id=provider_data['client_id'],
            client_secret=encrypt_data(provider_data['client_secret']),
            redirect_uri=provider_data['redirect_uri'],
            scopes=provider_data.get('scopes'),
            is_active=provider_data.get('is_active', True),
            auto_create_users=provider_data.get('auto_create_users', False),
            default_role=provider_data.get('default_role')
        )
        
        self.db.add(provider)
        self.db.commit()
        self.db.refresh(provider)
        
        return provider
    
    def get_oauth_providers(self, company_id: UUID, active_only: bool = True) -> List[OAuthProvider]:
        """Get Oauth Providers."""
        """Get OAuth providers for a company."""
        query = self.db.query(OAuthProvider).filter(
            OAuthProvider.company_id == company_id
        )
        
        if active_only:
            query = query.filter(OAuthProvider.is_active == True)
        
        return query.all()
    
    def get_login_attempts(
        self,
        company_id: Optional[UUID] = None,
        email: Optional[str] = None,
        limit: int = 100
    ) -> List[CompanyLoginAttempt]:
        """Get Login Attempts."""
        """Get login attempts with optional filters."""
        query = self.db.query(CompanyLoginAttempt)
        
        if company_id:
            query = query.filter(CompanyLoginAttempt.company_id == company_id)
        
        if email:
            query = query.filter(CompanyLoginAttempt.email == email)
        
        return query.order_by(desc(CompanyLoginAttempt.created_at)).limit(limit).all()
    
    def cleanup_expired_sessions(self):
        """Cleanup Expired Sessions."""
        """Clean up expired sessions."""
        expired_sessions = self.db.query(TenantSession).filter(
            and_(
                TenantSession.status == SessionStatus.ACTIVE,
                TenantSession.expires_at < datetime.utcnow()
            )
        ).all()
        
        for session in expired_sessions:
            session.status = SessionStatus.EXPIRED
        
        self.db.commit()
        return len(expired_sessions)
    
    def _enforce_session_limit(self, user_id: UUID, company_id: UUID, limit: int):
        """ Enforce Session Limit."""
        """Enforce concurrent session limit."""
        active_sessions = self.db.query(TenantSession).filter(
            and_(
                TenantSession.user_id == user_id,
                TenantSession.company_id == company_id,
                TenantSession.status == SessionStatus.ACTIVE,
                TenantSession.expires_at > datetime.utcnow()
            )
        ).order_by(TenantSession.last_activity).all()
        
        if len(active_sessions) >= limit:
            # Terminate oldest sessions
            sessions_to_terminate = active_sessions[:len(active_sessions) - limit + 1]
            
            for session in sessions_to_terminate:
                session.status = SessionStatus.TERMINATED
                session.terminated_reason = "session_limit_exceeded"
            
            self.db.commit()